autode/opt/coordinates/primitives.py

```diff
@@ -0,0 +1,222 @@
+import numpy as np
+from abc import ABC, abstractmethod
+from typing import Tuple
+
+
+class Primitive(ABC):
+    """Primitive internal coordinate"""
+
+    @abstractmethod
+    def __call__(self,
+                 x: 'autode.opt.coordinates.CartesianCoordinates'
+                 ) -> float:
+        """Return the value of this PIC given a set of cartesian coordinates"""
+
+    @abstractmethod
+    def derivative(self,
+                   i:         int,
+                   component: str,
+                   x:         'autode.opt.coordinates.CartesianCoordinates'
+                   ) -> float:
+        r"""
+        Calculate the derivative with respect to a cartesian coordinate
+
+        .. math::
+
+            \frac{dq}
+                  {d\boldsymbol{X}_{i, k}} {\Bigg\rvert}_{X=X0}
+
+        where :math:`q` is the primitive coordinate and :math:`\boldsymbol{X}`
+        are the cartesian coordinates.
+
+        -----------------------------------------------------------------------
+        Arguments:
+            i: Cartesian index to take the derivative with respect to;
+                0-N for N atoms
+
+            component: Cartesian component (x, y, z) to take the derivative
+                       with respect to. {'x', 'y', 'z'}
+
+            x: Cartesian coordinates
+
+        Returns:
+            (float): Derivative
+        """
+
+    @staticmethod
+    def _cart_component_to_idx(component: str) -> int:
+        """Convert the name of a Cartesian component (or direction) to an
+        index used for array indexing"""
+
+        if component == 'x':
+            return 0
+
+        if component == 'y':
+            return 1
+
+        if component == 'z':
+            return 2
+
+        else:
+            raise ValueError(f'Cannot convert component {component} to an '
+                             f'index. Must be one of x, y, z')
+
+    @abstractmethod
+    def __eq__(self, other):
+        """Comparison of two primitive coordinates"""
+
+
+class _DistanceFunction(Primitive, ABC):
```

Comment:
Does this class need to inherit from `ABC`?

Reply:
I think so(?) as it doesn't implement the abstract methods of `Primitive`